File: algorithms/cpp/astar.cpp
#include <iostream>
#include <cmath>
#include <tuple>
#include <queue>
using namespace std;

// Calculează distanța Manhattan dintre două poziții din grilă
double distantaManhattan(int iPoz, int jPoz, int iStop, int jStop) {
    return abs(iPoz - iStop) + abs(jPoz - jStop);
}

// Algoritmul A* pentru a găsi un drum de la punctul de start la punctul de stop într-un labirint
void astar(int matrice[][10], int iDim, int jDim, int iStart, int jStart, int iStop, int jStop) {
    // Vectori care definesc mișcarea în cele 4 direcții: jos, dreapta, sus, stânga
    int di[4] = {1, 0, -1, 0};
    int dj[4] = {0, 1, 0, -1};

    priority_queue<tuple<int, int, int>> openSet; // Coada de priorități pentru noduri, ordonată după scorul f (inversat pentru a obține efect de min-heap)
    int gScore[iDim][jDim];  // Costul real de la start până la fiecare nod
    int fScore[iDim][jDim];  // Estimarea costului total de la start până la stop prin acel nod

    matrice[iStart][jStart] = 1;  // Marcare inițială a poziției de start (nu reprezintă scorul real)
    gScore[iStart][jStart] = 0;  // Costul de la start până aici este 0
    fScore[iStart][jStart] = distantaManhattan(iStart, jStart, iStop, jStop); // Estimarea distanței până la destinație
    openSet.push(make_tuple(-fScore[iStart][jStart], iStart, jStart)); // Se adaugă în coadă cu scorul negativ pentru min-heap

    while (!openSet.empty()) {
        int fScoreCurent, iPoz, jPoz;
        tie(fScoreCurent, iPoz, jPoz) = openSet.top();
        openSet.pop();
        fScoreCurent = -fScoreCurent; // Se obține valoarea pozitivă a scorului f extras din coadă

        // Se verifică toți cei 4 vecini ai nodului curent
        for (int k = 0; k < 4; k++) {
            int iVecin = iPoz + di[k];
            int jVecin = jPoz + dj[k];

            // Verifică dacă vecinul este în interiorul grilei și este un spațiu liber (0)
            if (iVecin >= 0 && iVecin < iDim && jVecin >= 0 && jVecin < jDim && matrice[iVecin][jVecin] == 0) {
                matrice[iVecin][jVecin] = matrice[iPoz][jPoz] + 1;  // Se actualizează matricea cu un "pas" în plus

                // Adaugă vecinul în coada de priorități pentru a fi procesat
                openSet.push(make_tuple(distantaManhattan(iVecin, jVecin, iStop, jStop), iVecin, jVecin));

                // Dacă s-a ajuns la destinație, se încheie căutarea
                if (iVecin == iStop && jVecin == jStop)
                    return;
            }
        }
    }
}

int main() {
    // Reprezentarea labirintului: 0 = spațiu liber, -1 = obstacol
    int matrice[6][10] = {
        { 0,  0,  0, -1, -1,  0,  0,  0, -1,  0},
        {-1, -1,  0, -1,  0, -1,  0,  0,  0, -1},
        {-1,  0,  0,  0,  0,  0,  0, -1, -1,  0},
        {-1, -1, -1,  0, -1, -1, -1,  0, -1,  0},
        {-1,  0,  0, -1, -1,  0, -1,  0,  0,  0},
        {-1,  0,  0, -1, -1,  0,  0,  0,  0,  0}
    };

    int iStart = 0, jStart = 0; // Coordonatele de pornire
    int iStop = 1, jStop = 8;   // Coordonatele destinației

    astar(matrice, 6, 10, iStart, jStart, iStop, jStop);

    // Afișează valoarea finală din poziția destinație (distanța parcursă dacă s-a ajuns acolo)
    cout << matrice[iStop][jStop];

    return 0;
}

File: algorithms/csharp/astar.cs
using System;
using System.Collections.Generic;

class Program {
    // Calculează distanța Manhattan dintre două puncte
    static int DistantaManhattan(int iPoz, int jPoz, int iStop, int jStop) {
        return Math.Abs(iPoz - iStop) + Math.Abs(jPoz - jStop);
    }

    // Funcția ASTAR pentru a găsi cel mai scurt drum într-un labirint
    static void AStar(int[,] matrice, int iStart, int jStart, int iStop, int jStop) {
        // Vectori de direcție: jos, dreapta, sus, stânga
        int[] di = { 1, 0, -1, 0 };
        int[] dj = { 0, 1, 0, -1 };

        var openSet = new PriorityQueue<(int i, int j), int>(); // Coada de priorități pentru parcurgerea pozițiilor
        int rows = matrice.GetLength(0);
        int cols = matrice.GetLength(1);
        int[,] gScore = new int[rows, cols];
        int[,] fScore = new int[rows, cols];

        matrice[iStart, jStart] = 1;  // Marcăm poziția de start cu distanța 1
        gScore[iStart, jStart] = 0;
        fScore[iStart, jStart] = DistantaManhattan(iStart, jStart, iStop, jStop);
        openSet.Enqueue((iStart, jStart), fScore[iStart, jStart]);

        while (openSet.Count > 0) {
            var (iPoz, jPoz) = openSet.Dequeue();

            // Verificăm fiecare dintre cei 4 vecini
            for (int k = 0; k < 4; k++) {
                int iVecin = iPoz + di[k];
                int jVecin = jPoz + dj[k];

                // Dacă vecinul este în interiorul grilei și este liber (0)
                if (iVecin >= 0 && iVecin < rows && jVecin >= 0 && jVecin < cols && matrice[iVecin, jVecin] == 0) {
                    matrice[iVecin, jVecin] = matrice[iPoz, jPoz] + 1;  // Incrementăm distanța

                    // Adăugăm vecinul în coadă pentru procesare ulterioară
                    openSet.Enqueue((iVecin, jVecin), DistantaManhattan(iVecin, jVecin, iStop, jStop));

                    // Dacă am ajuns la destinație, putem opri căutarea
                    if (iVecin == iStop && jVecin == jStop)
                        return;
                }
            }
        }
    }

    static void Main() {
        // Reprezentarea labirintului: 0 = liber, -1 = obstacol
        int[,] matrice = {
            { 0,  0,  0, -1, -1,  0,  0,  0, -1,  0},
            {-1, -1,  0, -1,  0, -1,  0,  0,  0, -1},
            {-1,  0,  0,  0,  0,  0,  0, -1, -1,  0},
            {-1, -1, -1,  0, -1, -1, -1,  0, -1,  0},
            {-1,  0,  0, -1, -1,  0, -1,  0,  0,  0},
            {-1,  0,  0, -1, -1,  0,  0,  0,  0,  0}
        };

        int iStart = 0, jStart = 0; // Coordonatele de start
        int iStop = 1, jStop = 8;   // Coordonatele de destinație

        AStar(matrice, iStart, jStart, iStop, jStop);

        // Afișăm distanța până la destinație (sau valoare implicită dacă nu s-a ajuns)
        Console.WriteLine(matrice[iStop, jStop]);
    }
}
